### DISCUSSION TITLE
How to run Aspire on multiple computers instead of running Aspire on Azure

### DISCUSSION THREAD

**Discussion by cysnet** [microsoft/aspire#1439] [Q&A] [answered]
How to run Aspire on multiple computers instead of running Aspire on Azure?

**Comment by cysnet** [comment]
Does Aspire only run in Azure?

**Reply by plppp2001** [reply]
Can I add Aspire to my Web API if my Prod web api app runs on IIS?

**Comment by DamianEdwards** [comment]
Aspire doesn't include a runtime control plane or hosting mechanism for outside of the developer scenario. It's intended the app is deployed to something like kubernetes or a container-based PaaS. Ultimately the apps are standard .NET projects that you can deploy and configure in any of the supported ways.

**Comment by HClausing** [comment]
I'm doing some tests here with the starter template, and it's working fine just with Kestrel.

**EDIT**: Published AppHost project seems to be the same as Debug one. `ASPNETCORE_ENVIRONMENT` variable on running projects still showing "Development" on Dashboard's Enviroment variables dialog, when running from apphost.exe file.

**Comment by cysnet** [comment]
@DamianEdwards Is there a document to describe how to deploy aspire applications to k8s?

**Comment by davidfowl** [comment]
There is no automated deployment to k8s support in aspire preview1. Deploying to k8s is manual right now. We plan to investigate this in the future.

**Comment by cysnet** [comment]
@davidfowl how to deploy aspire applications to k8s manually? create deployments for every aspire application?

**Comment by cysnet** [comment]
@davidfowl Will multi-virtual machine deployment be supported in the future instead of k8s?

**Comment by davidfowl** [comment]
> @davidfowl Will multi-virtual machine deployment be supported in the future instead of k8s?

No, never. We're not rebuilding Kubernetes. The orchestration here is optimized for development scenarios. If you want anything on multiple machines, then use k8s.

**Comment by cysnet** [comment]
@davidfowl When is it expected that there will be documentation on how to deploy through k8s? And before this, could I only deploy to Azure?

**Comment by davidfowl** [comment]
Support for automatically generating and deploying to k8s? Months away. Maybe 2 (this experience needs to be designed). Yes, you can deploy to azure using AZD. Aspire itself cannot deploy to azure, the azure tools understand aspire manifests.

If you want to help design the k8s experience follow and contribute to this issue https://github.com/dotnet/aspire/issues/830

**Comment by HClausing** [comment]
> > @davidfowl Will multi-virtual machine deployment be supported in the future instead of k8s?
> 
> No, never. We're not rebuilding Kubernetes. The orchestration here is optimized for development scenarios. If you want anything on multiple machines, then use k8s.

Hi @davidfowl , I'm not going against the plans. But not only me, but many users from communities that I participate liked a lot what we've seen on .NET Conf about the .NET Aspire. I guess it so much powerfull for just to stay only on development scenarios. It's very easy to setup and run just using Kestrel + YARP (when we don't need to use K8S neither Docker).

We'll very happy we could see .NET Aspire to running also on production scenarios.

I'm sorry if I understand something wrong.

**Comment by davidfowl** [comment]
I don't think there's anything wrong with your desire to want this experience for deployments to non-cloud environments, but the reason we target the cloud and things like k8s for deployment is because there's a standard set of APIs for interacting with and configuring the compute, networking and anything required to make these systems run.

If you want to deploy to another environment then you can use aspire's extensible system but you have to build a tool that understands your target environment. I can help guide you to build it, but it's not something we are building as part of aspire.

**Comment by HClausing** [comment]
> I don't think there's anything wrong with your desire to want this experience for deployments to non-cloud environments, but the reason we target the cloud and things like k8s for deployment is because there's a standard set of APIs for interacting with and configuring the compute, networking and anything required to make these systems run.
> 
> If you want to deploy to another environment then you can use aspire's extensible system but you have to build a tool that understands your target environment. I can help guide you to build it, but it's not something we are building as part of aspire.

Cool, thank you so much for the answer!

What I'll try to do is to deploy to a server using only Kestrel + Aspire + YARP. It's alread works fine, but I've noted that Publish option on VS Preview 17.9 is compiling the projects looks like the Debug configuration, and it doesn't  copy the compiled files to desired target folder (just for the \bin\Release).

After it, I can immagine for the first time an extension that can allow to restart some service from the list on dashboard, but just for authenticated admins (role).

Do you believe it can be possible? For sure, not from official Aspire package, since it's not on roadmap, but into another extension.

**Comment by davidfowl** [comment]
Are you running your applications in containers or just as executables on the machine? 

I find these discussions to be more useful with diagrams:

![image](https://github.com/dotnet/aspire/assets/95136/4cf0f925-54f3-43aa-8080-5900567a6533)

What features of aspire do you want to persist in your production environment?
- The dashboard 👍🏾 
- service discovery 👍🏾 
- The orchestrator (which is only usable in development)? If yes, what are you using today to run your applications in prod? 
- What else?

**Comment by HClausing** [comment]
> Are you running your applications in containers or just as executables on the machine?

Actually I pretend to run on the prodution environment without containers. I like the ideia of Aspire starting every depedant project or executable by itself and showing their status into a deshboard (And maybe on future ofering a restart option on dashboard or function .Restart(...); ).

> What features of aspire do you want to persist in your production environment?
> 
> * The dashboard 👍🏾
> * service discovery 👍🏾
> * The orchestrator (which is only usable in development)? If yes, what are you using today to run your applications in prod?
> * What else?

* The dashboard:  It's good to check the health/logs from every dependant project/executable (services / microservices);
* Service discovery: Service discovery: I didn't kwon this feature until aspire, and I liked it alot;
* The orchestrator: Maybe I misunderstood this feature, but nowadays I'm running every executable from required services by scheduled tasks on OS (triggering they from OS startup - not requires an user to log on). This feature will simplify it into a single scheduled task instead many. It can allow on-premises deployment, for example, without docker and k8s.

Based on diagram, without Aspire, and not using Docker, I should configure four scheduled tasks on OS, and sometimes one or more can exit and the related services will also fail, and we only discover by the Task Manager (if Windows Server) or Event Viewer. The .NET Aspire on this preview state already helps a lot.

**Comment by samsp-msft** [comment]
Maybe we should be looking at how to enable the projects for Aspire be more easily run as a windows service or linux daemon for non-cloud deployment scenarios? As long as the apps can be F5'd while in VS, the eventual hosting model shouldn't matter.

**Comment by davidfowl** [comment, marked as answer]
Aspire provides an orchestrator for development but not for production. If you want to use the aspire model in your deployed application then you need an orchestrator. I'd recommend Kubernetes.

> Based on diagram, without Aspire, and not using Docker, I should configure four scheduled tasks on OS, and sometimes one or more can exit and the related services will also fail, and we only discover by the Task Manager (if Windows Server) or Event Viewer. The .NET Aspire on this preview state already helps a lot.

If you go custom on prem, then you're pulling all of these pieces together on your own. To get the aspire experience on your windows server then you need to figure out:
- How to deploy the dashboard and where it runs
- How the dashboard gets information about the running apps and how it gets their information (this can be done but it would need to be written)
- How those services send telemetry to the dashboard so that you get telemetry.

All of this could be built by a sufficiently motivated set of people. The aspire team doesn't have interest in building this but would provide guidance on how you can make it work.

**Reply by aateeque** [reply]
Landed here from this [thread](https://x.com/aateeque/status/1766799289617899963?s=20) with yourself  wanted to explore this idea further...where to start looking into the orchestrator & dashboard? Want to tell it about running apps on a single machine...

**Reply by davidfowl** [reply]
To build an experience like this, you have to break down aspire's architecture. The first thing I can do is explain what happens when you run locally today:

![image](https://github.com/dotnet/aspire/assets/95136/830cd44d-5298-4b15-8c0f-2e999d3606f0)

**Reply by davidfowl** [reply]
### Local Orchestrator

When developing locally, `Aspire.Hosting` builds up an object model of "resources", these can be containers, executables, projects, parameters or it can be completely custom (you can model a custom resource object) and it submits requests to this local orchestrator called DCP (The Developer Control Plane).

DCP is the process and container orchestrator that is optimized for local development (talks to IDEs, handles the debugger protocols etc). This component isn't supported in deployed environments. It handles launching processes (either via the IDE or not), it launches the containers using the container runtime (docker or podman), and launches containers and handles networking between them. This is all modeled in C# using Aspire.Hosting (DCP is an implementation detail).

### Dashboard

**Aspire.Hosting** is responsible for configuring the dashboard and setting up projects to push telemetry to the right endpoint. It also tells the DCP to launch the dashboard executable.

The dashboard has 2 functions, collecting telemetry (OTLP endpoint) and showing the resources list and console logs. You can run the dashboard standalone today using the container image (or as an executable. It comes with the aspire workload) and it’ll be usable and an otlp endpoint.

Making the list of resources show up in the dashboard requires a *resource server*. This is a gRPC API that the dashboard queries to understand the state of resources and show basic logs for each resource. When running locally, the resource server is in Aspire.Hosting. It queries the local state and exposes it to the resource sever.

### The Deployment Manifest

There's a deployment manifest that you can output from the apphost that is a serialized version of the application model (what we call the apphost project). You can work on translating this manifest into whatever is required to deploy to your target environment.

### Single Box deployments

One of the main things to figure out for single box deployments are the experience you want to build. What is it you want to build to make this end-to-end work. 

1. How do these services get to the target machine? 
1. What are you going to use for orchestration? What is launching your processes/containers on the target machine?
1. How does the dashboard work in that environment? 
    - You can get telemetry just fine by configuring applications the right way (setting up the right OTLP endpoint).
    - What is the resource server in this case? How does it tell the dashboard what is running?

PS: This is why it's easy to target systems like Kubernetes instead of bespoke on prem environments. There's a programmatic API for provisioning resources etc.

### A starting point

Here's a starting point setup to answer some of these questions for running on a windows server.

> How do these services get to the target machine

Don't automate this, do this however you do it today, ftp, copying bits over manually whatever works. You need to be able to map these bits to what configured in the original app host (essentially being able to map it to the manifest). Now you have your dotnet published applications on the machine somehow and you know how they map to projects in the manifest.

> What are you going to use for orchestration? What is launching your processes/containers on the target machine?

Build a bespoke orchestrator!

Make a program that processes the manifest and runs each of the applications as a windows service, including the dashboard executable (after installing the workload). This program also needs to allocate and assign random ports (or allocate ports in general) and assign them to each process using the right configuration. You'll likely need to produce a configuration file (like another appsettings file) with the configuration from the manifest expressed as environment variables (I'm not sure how easy it is to set per process env variables from windows services).

> What is the resource server in this case? How does it tell the dashboard what is running?

Your orchestrator that launched those windows services will keep running and query the windows services for health and report the status to the dashboard via the gRPC API.

**PS: We want to make it this possible for Aspire.Hosting to target different orchestration runtimes (this work is underway) but it's not fully there yet.**

### Use Aspir8 as a starting point or contribute to it

[Aspire8](https://github.com/prom3theu5/aspirational-manifests) has code to parse the manifest (written in C#) and generate k8s yaml or docker compose files. I'm sure it can be extended to enable something that runs on a single box but [there are still missing pieces around the dashboard and resource server](https://github.com/prom3theu5/aspirational-manifests/issues/118) that could be contributed and reused here.

**Reply by HClausing** [reply]
Thanks for this amazing tutorial, however, I realized that on preview 4 it's not possible to publish the host project to local folder anymore:

![image](https://github.com/dotnet/aspire/assets/15158923/cd89ef07-3431-4cd7-a5b3-ccb4c1d5387e)

So, I'm not able to test the Starting Point tips (except by copying files from \bin\debug\net8.0, but I won't get the assemblies compiled to release) , how it's described in "_How do these services get to the target machine section_".

I've realized that if I run "AspireApp1.AppHost.exe" (even if I copy bins to another folder), it runs all another related projects also, like the Visual Studio debugger does. I guess it stores the soluction path in somewhere. Maybe it won't work if I copy they to another device.

![image](https://github.com/dotnet/aspire/assets/15158923/f85e25bb-4f2d-46e1-95a2-ca255150d882)

![image](https://github.com/dotnet/aspire/assets/15158923/d1ead01d-4668-4de2-9f6c-2a0f72d546ab)

Now I'll study more about the "Build a bespoke orchestrator!" section. I'm starting to understand that on target machine we won't actually run the AppHost executable, and just our own build orchestrator.

**Reply by HClausing** [reply]
After some tryings, I've figured a good option: Use of `builder.AddExecutable()`, and manually set the OTEL_EXPORTER_OTLP_ENDPOINT variable on executable's appsettings.json. I guess it's a good way to use Aspire on premise deploy.

Now I just looking for a way to remove the "unknown_service:" prefix from Dashboard naming.

![image](https://github.com/dotnet/aspire/assets/15158923/fbd97401-672c-4469-8211-ef95f5665c8e)

![image](https://github.com/dotnet/aspire/assets/15158923/8023b625-3ad5-4067-9c86-f46187c018b5)

![image](https://github.com/dotnet/aspire/assets/15158923/ebe92ddf-bf80-42d5-95d2-0b87dbf6765d)

![image](https://github.com/dotnet/aspire/assets/15158923/d8876f76-6fbd-4dfa-bd75-6c32ba8175b1)

![image](https://github.com/dotnet/aspire/assets/15158923/c4d3e024-3bce-4eb7-9bbd-3a96b84d5381)

**Reply by davidfowl** [reply]
> After some tryings, I've figured a good option: Use of builder.AddExecutable(), and manually set the OTEL_EXPORTER_OTLP_ENDPOINT variable on executable's appsettings.json. I guess it's a good way to use Aspire on premise deploy.

The built in orchestrator isn't supported outside of local developer environments.

**Reply by HClausing** [reply]
Sure. No problem at all. I'm just doing some experiments.

Even thinking on local developer environments, I didn't figure how to remove the "unknown_service:" prefix from Executables references on "Structured Logs" and "Metrics" page:

![image](https://github.com/dotnet/aspire/assets/15158923/a7026024-67e2-4edb-9014-279d0af48c06)

**Reply by davidfowl** [reply]
When you add a project we add environment variables that configure the OTLP exporter:

https://github.com/dotnet/aspire/blob/9073e0788b830aafa1e0dede786e861143e1a8b9/src/Aspire.Hosting/Otlp/OtlpConfigurationExtensions.cs#L68

**Reply by HClausing** [reply]
Nice, when adding `.WithOtlpExporter()` after `.AddExecutable()` solved well.

Thanks again @davidfowl ! Sorry if i'm bothering with many questions.

**Comment by HClausing** [comment]
Hi all,

About the standalone Dashboard: I've setup an test environment with success, but I've missed the "Resources" tab, because there's no URL for "DOTNET_RESOURCE_SERVICE_ENDPOINT_URL" (structures logs, metrics are all working fine with another exe's running in kestrel). 

This resource service, by the way, comes from Aspire.Hosting, right? Is it already expected by design? So, using it as standalone (I've clonned the repository and debug the Dashboard) we won't be able to se any Resource.

**Reply by davidfowl** [reply]
The resource server in Aspire.Hosting is for development purposes only. You'll need to write a new resource server for your environment.

**Comment by clarkezone** [comment]
Any updates on documentation for how to deploy to local k8s?  Do i still need to make my own manifests?

**Reply by davidfowl** [reply]
See https://github.com/prom3theu5/aspirational-manifests

**Comment by SeidChr** [comment]
I would relly love to see a soulution for environment where azre or k8s is just no valid solution. The companies i work with, for the most part, are relying on windows servers with IIS hosting still.
Is there any way we can benefit from aspire? Currently we use powershell scripts to deploy the software, which has to run on each machine on the cluster

**Reply by davidfowl** [reply]
You can use the dashboard if no matter where you’re hosted (run it as a container). You won’t benefit from aspire’s deployment if your target environment is an IIS server as there’s nothing coming out of the box for those scenarios (nor do we have plans to build something there)

**Reply by SeidChr** [reply]
Dashboard requires a container to run?

**Reply by davidfowl** [reply]
That's the easiest way to run it. The other way is to install the aspire workload and run the dashboard executable directly.

**Reply by SeidChr** [reply]
I would need a way to deploy it along with all other apps of the bundle into an iis environment.
Which means: no docker, no workloads. Just an IIS application folder on a plain old windows server.

I guess we will experiment writing a powershell allowing us to install directly from a manifest on IIS.

**Reply by davidfowl** [reply]
This isn't made to run on IIS, I'm not sure what your options are if you are constrained in that way.

**Reply by SeidChr** [reply]
What is it made for to run on?
I mean what are the system requirements to run the dashboard outside a container? When i got it right, it runs on my local mashine. Is it docker only?
Is it limited to linux for some reason?
Is there something which prohibits it to run out of process on windows?
If you dont know in detail, any potential risks will help out as well!

**Reply by davidfowl** [reply]
The dashboard is runnable as a standalone process or as a docker container. It runs on any OS that supports .NET (I think I mentioned that you can run it after you install the workload on that OS).

> I mean what are the system requirements to run the dashboard outside a container? When i got it right, it runs on my local mashine. Is it docker only?

.NET 8.

> Is there something which prohibits it to run out of process on windows?

You can't run it on IIS but it should run on windows just fine.

> If you dont know in detail, any potential risks will help out as well!

There's a documentation on the dashboard (pay attention to the security considerations)

https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/dashboard/overview?tabs=bash

**Reply by SeidChr** [reply]
The manifest looks promising, for a custom deployment. But please correct me if my assumption is false somehow: the Dashboard, or AppHost is not actually part of the manifest, and thus a feature of the engine which deploys the Application. 
Thus i have to include it in my deliverable asset, already foreseeing that the deployment engine i will use in my on prem environment (a custom made script of some sort) also deploys the dashboard?

**Reply by davidfowl** [reply]
> the Dashboard, or AppHost is not actually part of the manifest, and thus a feature of the engine which deploys the Application.

That's correct. It's up to the deployment tool to do what is necessary to deploy the dashboard (or turn the dashboard on if it already exists in the target environment).

> Thus i have to include it in my deliverable asset, already foreseeing that the deployment engine i will use in my on prem environment (a custom made script of some sort) also deploys the dashboard?

Sounds about right.

**Reply by SeidChr** [reply]
Ok, well, that breaks my illusion.
I was hoping to find a full description of the services which will be deployed in the template itself, so I can write a basic processor creating the different resources.
Having to deploy a service which is not in the manifest makes the whole idea of using the manifest as a basis hard to hold up.
I guess we rather include the dashboard in our classic deployment

**Reply by davidfowl** [reply]
Take a look at https://github.com/prom3theu5/aspirational-manifests. It's the most complete C# based tool that processes the aspire manifests to create docker compose, kubernetes manifest and helm manifests.

**Reply by aateeque** [reply]
> That's correct. It's up to the deployment tool to do what is necessary to deploy the dashboard (or turn the dashboard on if it already exists in the target environment).
> 
> >

Is Aspire Dashboard only packaged up into a container image? or is it available to run standalone outside of a container image? if so...where...there's [this](https://www.nuget.org/packages/Aspire.Dashboard/8.0.0-preview.2.23619.3) which seems outdated?

**Reply by davidfowl** [reply]
It's packaged as a container image and as a part of the workload (using these packages) https://www.nuget.org/packages/Aspire.Dashboard.Sdk.win-x64/8.0.1 (there are architecture specific packages). These aren't normal nuget packages that you install as part of your projects. They are standalone executables.

**Reply by HClausing** [reply]
> It's packaged as a container image and as a part of the workload (using these packages) https://www.nuget.org/packages/Aspire.Dashboard.Sdk.win-x64/8.0.1 (there are architecture specific packages). These aren't normal nuget packages that you install as part of your projects. They are standalone executables.

I'm thinking on scenario that can be usefull do deploy on IIS / Kestrel, please let me know if I'm not wrong:

1) Build a basic console/asp.net project that is responsible to run Aspire.Dashboard and many others services/projects (like Aspire does in development, but without the Resources, etc);

2) Deploy the package content (executables) and many other published services at some `item 1` subfolder/relative path;

3) Create a scheduled task / daemon / service to run `item 1`.

**Reply by SeidChr** [reply]
I've managed to extract the runtime and actually run it on IIS. So there is a way! :)
EDIT: there is even a web.config in the folder, so to set up IIS it was only required to install the latest dotnet hosting bundle, add a single environment variable to the web-config disabling the login token and also configure the ports to be used in the appsettings and iis bindings.

```json
{
    "AllowedHosts": "*",
    "Dashboard": {
        "Frontend": {
            "EndpointUrls": "http://localhost:27524"
        },
        "Otlp": {
            "EndpointUrl": "http://localhost:27525"
        }
    }
}
```

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <location path="." inheritInChildApplications="false">
    <system.webServer>
      <handlers>
        <add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModuleV2" resourceType="Unspecified" />
      </handlers>
      <aspNetCore processPath=".\Aspire.Dashboard.exe" stdoutLogEnabled="false" stdoutLogFile=".\logs" hostingModel="inprocess">
        <environmentVariables>
          <environmentVariable name="DOTNET_DASHBOARD_UNSECURED_ALLOW_ANONYMOUS" value="true" />
        </environmentVariables>
      </aspNetCore>
    </system.webServer>
  </location>
</configuration>
```

**Reply by SeidChr** [reply]
@davidfowl 
I just managed to send telemetry to the standalone dashboard. Data is displayed in structured logs, metrics and traces 🎉.

The standard landing page of the dashboard would showing the services running in a typical aspire environment.
Running the dashboard standalone seems to disable this page along with the console output.
Yet showing the status and health of the services that are sending data is one of the more valueable functionalities I am looking for. (Console doesn't matter. The structured logs are already great!)

Before I dig into the dashboard code, can you point me to a documentation about what requirement needs to be fulfilled to display at least the service overview, so i can see running services?

In the end i need a nice way to display connection status information between the services. I am not sure if this would be a metric or displayable on that main page.

**Reply by davidfowl** [reply]
> I just managed to send telemetry to the standalone dashboard. Data is displayed in structured logs, metrics and traces 🎉.

Excellent!

> The standard landing page of the dashboard would showing the services running in a typical aspire environment.
> Running the dashboard standalone seems to disable this page along with the console output.

I answered this in detail https://github.com/dotnet/aspire/discussions/1439#discussioncomment-8737953

**Reply by SeidChr** [reply]
So basically i need to create "a resource server" which takes some config file describing the different services and then fetches the data from the services and makes it queryable for the dashboard.
Essentially a monitoring application.

**Comment by lpduyplus** [comment]
I found code can create resoure outside bebug as below. But i want to confuge run app in enviroment production, can yout help me!
step1: in web app project   name "client" (in my sample) right click and select  Add> >.NET Aspire Orchestrator support, it will be have three new project
+client.ServiceDefaults : it easy to copy to client project as extentions (it is content some method to generate endpoint and resource) then can remove it.
+client -> program.cs automatic add some code : 
`using client;
using client.Controllers;
using client.Models;
using client.Hubs;
var builder = WebApplication.CreateBuilder(args);
builder.AddServiceDefaults();           < -- here
...
var app = builder.Build();
app.MapDefaultEndpoints(); <-- and here
...
try
{
	app.Run();
}
catch (Exception ex)
{
	Console.WriteLine($"{ex.Message}\n{ex.StackTrace}");
}`
+client.AppHost

`**IProjects.cs**`
`using Microsoft.AspNetCore.Builder;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
namespace Projects;

[global::System.CodeDom.Compiler.GeneratedCode("Aspire.Hosting", null)]
[global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage(Justification = "Generated code.")]
[global::System.Diagnostics.DebuggerDisplay("Type = {GetType().Name,nq}, ProjectPath = {ProjectPath}")]
public class Client : global::Aspire.Hosting.IProjectMetadata
{

	public string ProjectPath => """ full path to \client.csproj""";
}
`
and fix some code on **programe.cs** to run localhost http after build release AppHost (run direction on window enviroment
if (string.IsNullOrEmpty(Environment.GetEnvironmentVariable("ASPNETCORE_URLS")))
{	
	Environment.SetEnvironmentVariable("ASPNETCORE_URLS", "http://localhost:5001");
}
var otlpEndpoint = Environment.GetEnvironmentVariable("DOTNET_DASHBOARD_OTLP_ENDPOINT_URL");
var otlpHttpEndpoint = Environment.GetEnvironmentVariable("DOTNET_DASHBOARD_OTLP_HTTP_ENDPOINT_URL");
Environment.SetEnvironmentVariable("ASPIRE_ALLOW_UNSECURED_TRANSPORT", "true");
if (!string.IsNullOrEmpty(otlpEndpoint) && !string.IsNullOrEmpty(otlpHttpEndpoint))
{
	DOTNET_DASHBOARD_OTLP_HTTP_ENDPOINT_URL.
	Environment.SetEnvironmentVariable("DOTNET_DASHBOARD_OTLP_HTTP_ENDPOINT_URL", null);
}
else if (string.IsNullOrEmpty(otlpEndpoint) && string.IsNullOrEmpty(otlpHttpEndpoint))
{	
	Environment.SetEnvironmentVariable("DOTNET_DASHBOARD_OTLP_ENDPOINT_URL", "http://localhost:4317");
}
var builder = DistributedApplication.CreateBuilder(args);

builder.AddProject<Projects.Client>("client");

builder.Build().Run();
